File: Distributed Computing Lab/Experiment-7/ChandyMisraHaas.java
/**
 * Name: Amey Mahendra Thakur
 * Course: Distributed Computing Lab (CSL802)
 * Roll No: 50 | Batch: B3
 * Date of Experiment: February 25, 2022
 * Repository: https://github.com/Amey-Thakur/DISTRIBUTED-COMPUTING-AND-DISTRIBUTED-COMPUTING-LAB
 * Description: Experiment 7 - Unified implementation of Chandy-Misra-Haas Deadlock Detection Algorithm (Deadlock & No-Deadlock scenarios).
 */

import java.util.Scanner;

public class ChandyMisraHaas {
    private static int numProcesses = 5;
    private static int[][] waitMatrix;
    private static boolean deadlockDetected = false;

    /**
     * Entry point for the Chandy-Misra-Haas Deadlock Detection Simulator.
     */
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("=" + "=".repeat(63));
        System.out.println(" CHANDY-MISRA-HAAS DISTRIBUTED DEADLOCK DETECTION SIMULATOR (JAVA) ");
        System.out.println("=" + "=".repeat(63));
        
        // 1. Configuration Selection
        System.out.println("\nSelect Experimental Configuration:");
        System.out.println("  1. Scenario A: Deadlock Case (Cycle: P1->P2->P3->P4->P1)");
        System.out.println("  2. Scenario B: No-Deadlock Case (Acyclic Graph)");
        System.out.print("\n[?] Enter choice (1/2): ");
        int choice = scanner.nextInt();

        if (choice != 1 && choice != 2) {
            System.out.println("[!] Invalid selection. Terminating.");
            scanner.close();
            return;
        }

        initializeWaitGraph(choice);

        // 2. WFG Visualization
        displayWaitGraph();

        // 3. Probe Initiation
        System.out.print("\n[?] Enter site ID initiating detection (1-" + numProcesses + "): ");
        int initiator = scanner.nextInt();

        if (initiator >= 1 && initiator <= numProcesses) {
            startDetection(initiator);
        } else {
            System.out.println("[!] Error: Site ID must be between 1 and " + numProcesses);
        }

        System.out.println("\n" + "=".repeat(65));
        scanner.close();
    }

    /**
     * Initializes the Adjacency Matrix for the Wait-For-Graph.
     */
    private static void initializeWaitGraph(int scenario) {
        waitMatrix = new int[numProcesses][numProcesses];
        
        // Common dependencies in both scenarios
        waitMatrix[0][1] = 1; // P1 depends on P2
        waitMatrix[1][2] = 1; // P2 depends on P3
        waitMatrix[2][3] = 1; // P3 depends on P4
        waitMatrix[2][4] = 1; // P3 depends on P5

        if (scenario == 1) {
            waitMatrix[3][0] = 1; // P4 depends on P1 (Closing the cycle: 1->2->3->4->1)
            System.out.println("\n[*] STATUS: Deadlock Scenario loaded.");
        } else {
            System.out.println("\n[*] STATUS: No-Deadlock Scenario loaded.");
        }
    }

    /**
     * Visualizes the current dependency state.
     */
    private static void displayWaitGraph() {
        System.out.println("\n[*] Wait-For-Graph Adjacency Matrix:");
        System.out.println("     P1  P2  P3  P4  P5");
        for (int i = 0; i < numProcesses; i++) {
            System.out.print("P" + (i + 1) + "  ");
            for (int j = 0; j < numProcesses; j++) {
                System.out.print(waitMatrix[i][j] + "   ");
            }
            System.out.println();
        }
    }

    /**
     * Facilitates the edge-chasing probe mechanism.
     */
    private static void startDetection(int initiatorSite) {
        int initiatorIdx = initiatorSite - 1;
        deadlockDetected = false;

        System.out.println("\n[*] Site S" + initiatorSite + " initiating deadlock detection probe...");
        
        for (int target = 0; target < numProcesses; target++) {
            if (waitMatrix[initiatorIdx][target] == 1) {
                System.out.println("    -> Sending Probe: (Init:" + initiatorSite + ", Sender:" + initiatorSite + ", Receiver:" + (target + 1) + ")");
                propagateProbe(initiatorIdx, target);
            }
        }

        if (deadlockDetected) {
            System.out.println("\n[RESULT] CRITICAL: DISTRIBUTED DEADLOCK DETECTED at Site S" + initiatorSite);
        } else {
            System.out.println("\n[RESULT] Verification Complete: No deadlock detected from site S" + initiatorSite);
        }
    }

    /**
     * Recursive probe propagation through the WFG.
     */
    private static void propagateProbe(int initiatorIdx, int currentIdx) {
        if (deadlockDetected) return;

        for (int next = 0; next < numProcesses; next++) {
            if (waitMatrix[currentIdx][next] == 1) {
                // Deadlock Condition: Probe returns to the initiating site
                if (next == initiatorIdx) {
                    System.out.println("    !! Probe returned to Initiator P" + (initiatorIdx + 1) + " !!");
                    System.out.println("    !! Path: (" + (initiatorIdx + 1) + ", " + (currentIdx + 1) + ", " + (next + 1) + ") --------> CYCLE DETECTED");
                    deadlockDetected = true;
                    return;
                }

                System.out.println("    -> Forwarding Probe: (Init:" + (initiatorIdx + 1) + ", Sender:" + (currentIdx + 1) + ", Receiver:" + (next + 1) + ")");
                propagateProbe(initiatorIdx, next);
            }
        }
    }
}

File: Distributed Computing Lab/Experiment-7/ChandyMisraHaas.py
# Name: Amey Mahendra Thakur
# Course: Distributed Computing Lab (CSL802)
# Roll No: 50 | Batch: B3
# Date of Experiment: February 25, 2022
# Repository: https://github.com/Amey-Thakur/DISTRIBUTED-COMPUTING-AND-DISTRIBUTED-COMPUTING-LAB
# Description: Experiment 7 - Unified implementation of Chandy-Misra-Haas Deadlock Detection Algorithm (Deadlock & No-Deadlock scenarios).

class ChandyMisraHaas:
    """
    Implements the Chandy-Misra-Haas edge-chasing algorithm.
    This algorithm detects distributed deadlock by sending probes along the Wait-For-Graph.
    """
    def __init__(self, n):
        self.n = n
        self.deadlock_found = False

    def reset(self):
        self.deadlock_found = False

    def initiate_probe(self, matrix, initiator):
        """
        Starts the probe-sending process from a detected wait condition.
        """
        self.reset()
        idx = initiator - 1
        print(f"\n[*] Site S{initiator} initiating deadlock detection probe...")
        
        for next_node in range(self.n):
            if matrix[idx][next_node] == 1:
                print(f"    -> Probe Sent: (Init:{initiator}, Sender:{initiator}, Receiver:{next_node+1})")
                self.propagate(matrix, idx, next_node)
        
        if not self.deadlock_found:
            print(f"\n[RESULT] Verification Successful: No deadlock detected from initiator S{initiator}.")
        else:
            print(f"\n[RESULT] CRITICAL: DISTRIBUTED DEADLOCK DETECTED (Cycle detected at Site S{initiator}).")

    def propagate(self, matrix, initiator_idx, current_idx):
        """
        Recursively propagates the probe (Initiator, Sender, Receiver).
        """
        if self.deadlock_found:
            return

        for k in range(self.n):
            if matrix[current_idx][k] == 1:
                # If probe returns to initiator, a cycle exists
                if k == initiator_idx:
                    print(f"    !! Deadlock: Probe (Init:{initiator_idx+1}, Sender:{current_idx+1}, Receiver:{k+1}) returned to initiator !!")
                    self.deadlock_found = True
                    return
                
                print(f"    -> Probe Forwarded: (Init:{initiator_idx+1}, Sender:{current_idx+1}, Receiver:{k+1})")
                self.propagate(matrix, initiator_idx, k)

def main():
    print("="*65)
    print(" CHANDY-MISRA-HAAS DISTRIBUTED DEADLOCK DETECTION SIMULATOR ")
    print("="*65)

    scenarios = {
        "1": {
            "name": "Scenario A: Deadlock Case (Cycle: P1->P2->P3->P4->P1)",
            "matrix": [
                [0, 1, 0, 0, 0], 
                [0, 0, 1, 0, 0], 
                [0, 0, 0, 1, 1], 
                [1, 0, 0, 0, 0], 
                [0, 0, 0, 0, 0]
            ]
        },
        "2": {
            "name": "Scenario B: No-Deadlock Case (Acyclic Graph)",
            "matrix": [
                [0, 1, 0, 0, 0], 
                [0, 0, 1, 0, 0], 
                [0, 0, 0, 1, 1], 
                [0, 0, 0, 0, 0], 
                [0, 0, 0, 0, 0]
            ]
        }
    }

    print("\nSelect Experimental Configuration:")
    for k, v in scenarios.items():
        print(f"  {k}. {v['name']}")
    
    choice = input("\n[?] Enter choice (1/2): ")
    if choice not in scenarios:
        print("[!] Invalid selection.")
        return

    matrix = scenarios[choice]["matrix"]
    
    print("\n[*] Wait-For-Graph Adjacency Matrix:")
    print("     P1  P2  P3  P4  P5")
    for i, row in enumerate(matrix):
        print(f"P{i+1}  " + "   ".join(map(str, row)))

    try:
        initiator = int(input("\n[?] Enter site initiating detection (1-5): "))
        if 1 <= initiator <= 5:
            detector = ChandyMisraHaas(5)
            detector.initiate_probe(matrix, initiator)
        else:
            print("[!] Site number must be between 1 and 5.")
    except ValueError:
        print("[!] Error: Invalid numeric input.")

    print("\n" + "="*65)

if __name__ == "__main__":
    main()
